File: code/workflow/decompose_kmeans.R
#!/usr/bin/Rscript

# Load libraries
library(tidyverse)
library(glue)
library(RMINC)
library(RcppCNPy)

###########################
# Arguments
###########################

# Testing arguments
# outdir <- "data/outputs/isocortex_thalamus_left"
# decompdir <- "data/outputs/isocortex_thalamus_left/decomposition_z/kmeans/10_clusters"
# n_clusters <- 10

# Read arguments
args <- commandArgs(trailingOnly = TRUE)
outdir <- args[1]
decompdir <- args[2]
n_clusters <- as.integer(args[3])

# Input files
mask_file_thalamus <- glue("{outdir}/thalamus_mask_highres.mnc")
mask_file_cortex <- glue("{outdir}/cortex_mask_lowres.mnc")
array_file_labels <- glue("{decompdir}/cluster_labels.npy")
array_file_centers <- glue("{decompdir}/cluster_centers.npy")


###########################
# Read data
###########################

print(glue("[{Sys.time()}] Reading data"))

mask_vol_thalamus <- mincGetVolume(mask_file_thalamus)
mask_vol_cortex <- mincGetVolume(mask_file_cortex)

p_thal <- length(mask_vol_thalamus[mask_vol_thalamus > 0.5])
p_cortex <- length(mask_vol_cortex[mask_vol_cortex > 0.5])

k_labels <- npyLoad(array_file_labels)
k_centers <- npyLoad(array_file_centers)

###########################
# Write out
###########################

# Link averages
file.symlink(normalizePath("data/resources/average_template_50.mnc"), glue("{decompdir}/average_template_50.mnc"))
file.symlink(normalizePath("data/resources/average_template_200.mnc"), glue("{decompdir}/average_template_200.mnc"))
file.symlink(normalizePath("data/resources/AMBA_relabeled_50um_right_side.mnc"), glue("{decompdir}/AMBA_relabeled_50um_right_side.mnc"))
file.symlink(normalizePath("data/resources/AMBA_relabeled_50um_left_side.mnc"), glue("{decompdir}/AMBA_relabeled_50um_left_side.mnc"))
file.symlink(normalizePath("data/resources/AMBA_relabeled_50um_right_side_THonly.mnc"), glue("{decompdir}/AMBA_relabeled_50um_right_side_THonly.mnc"))
file.symlink(normalizePath("data/resources/AMBA_relabeled_50um_left_side_THonly.mnc"), glue("{decompdir}/AMBA_relabeled_50um_left_side_THonly.mnc"))

# Write out cluster labels
ovol <- mask_vol_thalamus
ovol[] <- 0
ovol[mask_vol_thalamus > 0.5] <- k_labels
mincWriteVolume(ovol, glue("{decompdir}/cluster_labels.mnc"))

# Merge cluster labels with AMBA labels
system(cmd <- glue("mincmath -2 -add {decompdir}/cluster_labels.mnc data/resources/AMBA_relabeled_50um_right_side.mnc {decompdir}/cluster_labels_with_AMBA_relabeled_50um.mnc"))

# Merge cluster labels with AMBA labels, TH only
system(cmd <- glue("mincmath -2 -add {decompdir}/cluster_labels.mnc data/resources/AMBA_relabeled_50um_right_side_THonly.mnc {decompdir}/cluster_labels_with_AMBA_relabeled_50um_THonly.mnc"))

# Write out individual masks
for (i in 1:n_clusters) {
  ovol <- mask_vol_thalamus
  ovol[] <- 0
  outvec <- rep(0, length.out=p_thal)
  outvec[which(k_labels==i)] <- 1
  ovol[mask_vol_thalamus > 0.5] <- outvec
  mincWriteVolume(ovol, glue("{decompdir}/cluster_mask_{i}.mnc"))
}

# Write out cluster centers
for (i in 1:n_clusters) {
  ovol <- mask_vol_cortex
  ovol[] <- 0
  ovol[mask_vol_cortex > 0.5] <- k_centers[i,]
  mincWriteVolume(ovol, glue("{decompdir}/cluster_center_{i}.mnc"))
}

# Subdivide cortex based on center max
ovol <- mask_vol_cortex
ovol[] <- 0
ovol[mask_vol_cortex > 0.5] <- apply(k_centers, 2, which.max)
mincWriteVolume(ovol, glue("{decompdir}/cluster_center_max.mnc"))

# Subdivide cortex based on center max, after scaling
ovol <- mask_vol_cortex
ovol[] <- 0
ovol[mask_vol_cortex > 0.5] <- apply(t(scale(t(k_centers))), 2, which.max)
mincWriteVolume(ovol, glue("{decompdir}/cluster_center_scaled_max.mnc"))

# Subdivide cortex based on center min
ovol <- mask_vol_cortex
ovol[] <- 0
ovol[mask_vol_cortex > 0.5] <- apply(k_centers, 2, which.min)
mincWriteVolume(ovol, glue("{decompdir}/cluster_center_min.mnc"))

# Subdivide cortex based on center min, after scaling
ovol <- mask_vol_cortex
ovol[] <- 0
ovol[mask_vol_cortex > 0.5] <- apply(t(scale(t(k_centers))), 2, which.min)
mincWriteVolume(ovol, glue("{decompdir}/cluster_center_scaled_min.mnc"))

File: code/workflow/decompose_kmeans.py
import numpy as np
import os
import sys
import pandas as pd
from datetime import datetime
from sklearn.cluster import KMeans
from pyminc.volumes.factory import *

# %% Script arguments
outdir = sys.argv[1]
decompdir = sys.argv[2]
matfile = sys.argv[3]
n_clusters = int(sys.argv[4])

# Testing arguments
# outdir = "data/outputs/isocortex_thalamus_left"
# decompdir = "data/outputs/isocortex_thalamus_left/decomposition_z/kmeans/10_clusters"
# matfile = "correlation_z.npy"
# n_clusters = 10

# %% Script specific calls

# KMeans: each elkan iteration in serial takes ~ 5 seconds
# 15 GB base + 15 GB per job
# Therefore running 5 jobs in parallel takes 90 GB, ~ 2 hours, for n_init=10, max_iter=500

model = KMeans(n_clusters=n_clusters, n_init=10, max_iter=500, verbose=1, n_jobs=5, algorithm="elkan", precompute_distances=False)
rscript = "code/workflow/decompose_kmeans.R"

# %% Create output directory
print("[{time}] Making output directory".format(time=datetime.now()))

os.makedirs(decompdir, exist_ok=True)


# %% Read data
print("[{time}] Reading data".format(time=datetime.now()))

mask_file_thalamus = os.path.join(outdir, "thalamus_mask_highres.mnc")
mask_file_cortex = os.path.join(outdir, "cortex_mask_lowres.mnc")
zmat_file = os.path.join(outdir, matfile)

mask_vol_thalamus = volumeFromFile(mask_file_thalamus, labels=True)
mask_vol_cortex = volumeFromFile(mask_file_cortex, labels=True)

p_thal = len(mask_vol_thalamus.data[mask_vol_thalamus.data > 0.5])
p_cortex = len(mask_vol_cortex.data[mask_vol_cortex.data > 0.5])

zmat = np.load(zmat_file)
zmat.shape

# %% Cluster/decompose
print("[{time}] Clustering/decomposing".format(time=datetime.now()))

model.fit(zmat)

print("[{time}] Done main computation".format(time=datetime.now()))

# %% Write out results
print("[{time}] Writing out npy arrays".format(time=datetime.now()))

np.save("{decompdir}/cluster_labels.npy".format(decompdir=decompdir), (model.labels_.astype('float64') + 1))
np.save("{decompdir}/cluster_centers.npy".format(decompdir=decompdir), model.cluster_centers_)

df = pd.DataFrame(data=model.get_params(), index=np.arange(0, 1))
df['inertia'] = model.inertia_
df.to_csv("{decompdir}/model_params.csv".format(decompdir=decompdir))

# %% Call Rscript to write out MNC files
print("[{time}] Writing out minc files via Rscript".format(time=datetime.now()))

cmd = "/usr/bin/time -v Rscript {rscript} {outdir} {decompdir} {n_clusters}".format(rscript=rscript, outdir=outdir, decompdir=decompdir, n_clusters=n_clusters)
os.system(cmd)

# %% Finally
print("[{time}] Done!".format(time=datetime.now()))
